src/librustc_codegen_llvm/base.rs

```diff
@@ -1283,6 +1283,40 @@ pub fn provide(providers: &mut Providers) {
             .unwrap_or_else(|| panic!("failed to find cgu with name {:?}", name))
     };
 
+    providers.backend_optimization_level = |tcx, cratenum| {
+        let for_speed = match tcx.sess.opts.optimize {
+            config::OptLevel::No => config::OptLevel::Default,
+            config::OptLevel::Less => config::OptLevel::Less,
+            config::OptLevel::Default => config::OptLevel::Default,
+            config::OptLevel::Aggressive => config::OptLevel::Aggressive,
+            config::OptLevel::Size => config::OptLevel::Default,
+            config::OptLevel::SizeMin => config::OptLevel::Default,
+
+        };
+
+        let mut need_speed = false;
+        let (defids, _) = tcx.collect_and_partition_mono_items(cratenum);
+        for id in &*defids {
+            let CodegenFnAttrs { optimize, .. } = tcx.codegen_fn_attrs(*id);
+            match optimize {
+                attr::OptimizeAttr::None => continue,
+                attr::OptimizeAttr::Size => continue,
+                attr::OptimizeAttr::Speed => {
+                    need_speed = true;
+                    break;
+                }
+            }
+        }
+        return if need_speed {
+            // If any of the items require optimize(speed), we must use speed optimisation globally
+            // and set either `optnone` or corresponding size-oriented optimisation level for the
+            // rest of the items in the crate. Those attributes are set specifically
+            for_speed
```

Comment:
I suspect that this will have some unintended fallout. While function passes are going to skip optnone functions, we'll also end up populating a full module pass manager, where likely not all module passes will ignore optnone functions (not to mention non-functions -- I don't believe there is an optnone for global variables.) Marking a single function for `optimize(speed)` may end up influencing codegen for the whole crate in non-trivial ways.

Reply:
This is a known caveat. Alas, to get `optimize(speed)` (or `optimize(speed)`, now that I think about it) to work at all, it is necessary to initialize some sort of pass manager.

Are you saying it is possible to have separate managers for functions and the rest of the module? That might be a better approach then.

Reply:
Alternative is to have these attributes to not work with `-Copt-level=0` at all, but that is not something I intended when I wrote the RFC.